tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetSetService.scala

```diff
@@ -17,4 +19,29 @@ class PublicAssetSetService @Inject()(val identityRepository: IdentityRepository
         case Seq() => SMono.just(identityIds)
         case noExistIdentityIds => SMono.error(PublicAssetIdentityIdNotFoundException(noExistIdentityIds))
       }
-}
\ No newline at end of file
+
+  def create(username: Username, creationRequest: PublicAssetCreationRequest): SMono[PublicAssetStorage] =
+    SMono(publicAssetRepository.create(username, creationRequest))
+      .onErrorResume {
+        case _: PublicAssetQuotaLimitExceededException => cleanUpPublicAsset(username)
+          .filter(cleanedUpSize => cleanedUpSize >= creationRequest.size.value)
+          .switchIfEmpty(SMono.error(new PublicAssetQuotaLimitExceededException))
+          .flatMap(_ => SMono(publicAssetRepository.create(username, creationRequest)))
+      }
+
+  // return the total size of public assets that was cleaned up
+  def cleanUpPublicAsset(username: Username): SMono[Long] =
+    SFlux(identityRepository.list(username))
+      .map(_.id)
+      .collectSeq()
+      .flatMapMany(existIdentityIds => listPublicAssetNeedToDelete(username, existIdentityIds))
+      .concatMap(publicAssetMetadata => SMono(publicAssetRepository.remove(username, publicAssetMetadata.id))
+        .`then`(SMono.just(publicAssetMetadata.size.value)), 1)
+      .collectSeq()
+      .map(_.sum)
```

Comment:
Ideally I'd like to have a sort by date to delete older public assets first as it will decrease inconvenience as it will impact older mails...
We may add a date field to public assets to that end IMO. If missing we can infer "now".

Reply:
The sort is unnecessary when we remove ALL unreference publicasset

Reply:
> The sort is unnecessary when we remove ALL unreference publicasset

I do not agree here.

Isn't this excessive?

Shouldn't we break only the oldest public assets? 

Being smaller in scope here means that we decrease the chance to impact emails sent with existing public assets...

Reply:
I updated in last commit 8855b0406c1e637b602865f0f6a7c55400541a6d with logic:
> We will cleanup each by each (after sorted) and stop when we got at least enough free size for new publicAsset

Reply:
> We will cleanup each by each (after sorted) and stop when we got at least enough free size for new publicAsset?

That was the idea.